### PR TITLE
Load testing with 1000 ServiceEntries and 512 pods

### PR DESCRIPTION
**Pull request by costinm**
The main fix is in external/servicediscovery.go - the yaml to proto conversion was so slow that pretty much everything was stopping, including the readiness /endpointz after restart. The root cause was that needsUpdate was not reset after update, and we haven't done tests with larger numbers of external 
services or VMs.

In the process I added a simpler /ready handler, and added some experimental env to throttle for large deployments ( where memory/CPU can spike a lot ). 

This PR also adds a lock to prevent 2 pushes from overlapping (causing invalid config), and to suppress a push if a more recent version is already in progress. Both changes can be disabled with an env variable in case something goes wrong - but should avoid some hard to debug corner cases.

### CODE REVIEW

File: pilot/pkg/proxy/envoy/v2/ads.go [hunk 1]
```diff
@@ -140,6 +144,12 @@ func init() {
 	prometheus.MustRegister(pushTimeouts)
 	prometheus.MustRegister(pushes)
 	prometheus.MustRegister(pushErrors)
+
+	// Experimental env to disable push suppression
+	// By default a pod will not receive a push from an older version if a
+	// push for a newer version has started. This behavior can be disabled
+	// for testing or in special cases (bugs or corner cases found, etc)
+	suppressPush = os.Getenv("PILOT_SKIP_SUPPRESS") == ""
```
Comment (rshriram) [inline pilot/pkg/proxy/envoy/v2/ads.go]: Allow_concurrent_push

File: pilot/pkg/proxy/envoy/v2/ads.go [hunk 2]
```diff
@@ -507,6 +520,13 @@ func (s *DiscoveryServer) IncrementalAggregatedResources(stream ads.AggregatedDi
 }
 
 func (s *DiscoveryServer) pushAll(con *XdsConnection, pushEv *XdsEvent) error {
+	// Prevent 2 overlapping pushes. Disabled if push suppression is disabled
+	// (as fail-safe in case of bugs)
+	if suppressPush {
```
Comment (rshriram) [inline pilot/pkg/proxy/envoy/v2/ads.go]: This doesn’t seem to match with the comment above where suppresspush was meant to disable the behavior. Did you mean !suppressPush? (Also for more clarity, might be better to call this as allowConcurrentPushes)

File: pilot/pkg/proxy/envoy/v2/ads.go [hunk 3]
```diff
@@ -517,7 +537,13 @@ func (s *DiscoveryServer) pushAll(con *XdsConnection, pushEv *XdsEvent) error {
 				time.Since(pushEv.push.Start), string(out))
 		}
 	}()
-	// TODO: check version, suppress if changed
+	// check version, suppress if changed.
+	currentVersion := versionInfo()
+	if suppressPush && pushEv.version != currentVersion {
```
Comment (rshriram) [inline pilot/pkg/proxy/envoy/v2/ads.go]: Pretty sure now that this is !allowConcurrentPush and that the env var you have above needs to be inverted. Ie false by default.

File: pilot/pkg/proxy/envoy/v2/ads.go [hunk 4]
```diff
@@ -609,7 +635,24 @@ func (s *DiscoveryServer) AdsPushAll(version string) {
...
+			adsLog.Warnf("Invalid push throttle %s", pushThrottleCountEnv)
+		}
+	}
+
+	i := 0
 	for _, c := range tmpMap {
+		i++
+		if pushThrottle > 0 && i%pushThrottle == 0 {
+			time.Sleep(100 * time.Millisecond)
+		}
```
Comment (rshriram) [inline pilot/pkg/proxy/envoy/v2/ads.go]: How is this going to help pilot deal with memory spikes , by delaying pushes to Envoy ?

File: pilot/pkg/proxy/envoy/v2/debug.go [hunk 5]
```diff
@@ -659,6 +661,15 @@ func writeAllADS(w io.Writer) {
 	fmt.Fprint(w, "]\n")
 }
 
+func (s *DiscoveryServer) ready(w http.ResponseWriter, req *http.Request) {
+	if s.ConfigController != nil {
+		if !s.ConfigController.HasSynced() {
+			w.WriteHeader(503)
```
Comment (rshriram) [inline pilot/pkg/proxy/envoy/v2/debug.go]: Where is hasSynced implemented ?

File: pilot/pkg/serviceregistry/external/servicediscovery.go [hunk 6]
```diff
@@ -235,30 +235,38 @@ func (d *ServiceEntryStore) update() {
...
 		}
 	}
+
+	d.storeMutex.Lock()
+	d.instances = di
+	d.ip2instance = dip
+	// Without this pilot will become very unstable even with few 100 ServiceEntry
+	// objects - the N_clusters * N_update generates too much garbage
+	// ( yaml to proto)
+	// This is reset on any change in ServiceEntries
```
Comment (rshriram) [inline pilot/pkg/serviceregistry/external/servicediscovery.go]: Nice find.

File: pilot/pkg/proxy/envoy/v2/ads.go (hunk 1 above)
Comment (costinm) [reply pilot/pkg/proxy/envoy/v2/ads.go]: done

Comment (costinm) [reply pilot/pkg/proxy/envoy/v2/ads.go]: Regarding charts - will change in separate PR, want to test the /ready a bit more.

File: pilot/pkg/proxy/envoy/v2/ads.go (hunk 2 above)
Comment (costinm) [reply pilot/pkg/proxy/envoy/v2/ads.go]: done

File: pilot/pkg/proxy/envoy/v2/ads.go (hunk 4 above)
Comment (costinm) [reply pilot/pkg/proxy/envoy/v2/ads.go]: Yes, right now on change we start pushing to all sidecars at once - and all start doing yaml-to-proto and heavy mem allocation. By sleeping after each 50 sidecars (for example) we reduce memory pressure a bit. 

Eventually, as discussed, we'll do smarter sequencing here, like wait for a set of canaries and add rate control.

File: pilot/pkg/proxy/envoy/v2/debug.go (hunk 5 above)
Comment (costinm) [reply pilot/pkg/proxy/envoy/v2/debug.go]: In k8s config controller - returns k8s 'has synced' state.  I think the others return true by default. 

It means k8s has read all configs  - so requests to pilot will have the needed state.

Note that endpoints and services might not be fully loaded - we would need to check the registry controller.

File: .circleci/config.yml [hunk 7]
```diff
@@ -19,10 +20,11 @@ integrationDefaults: &integrationDefaults
   machine: true
   working_directory: /go/src/istio.io/istio #same as docker
   environment:
-    CHANGE_MINIKUBE_NONE_USER: true
-    HUB: docker.io/dnerepo
-    TAG: dontpush
-    GOPATH: /go
+    - CHANGE_MINIKUBE_NONE_USER: true
+    - GOPATH: /go
+    - SKIP_CLEANUP: true
```
Comment (andraxylia) [inline .circleci/config.yml]: Do we want to skip cleanup?

**Pull request merged**